File: samples/client/3_1_0_unit_test/java/src/main/java/org/openapijsonschematools/client/schemas/validation/MinContainsValidator.java
package org.openapijsonschematools.client.schemas.validation;

import org.checkerframework.checker.nullness.qual.Nullable;
import org.openapijsonschematools.client.exceptions.ValidationException;

import java.util.List;

public class MinContainsValidator implements KeywordValidator {
    public final int minContains;

    public MinContainsValidator(int minContains) {
        this.minContains = minContains;
    }

    @Override
    public @Nullable PathToSchemasMap validate(
        ValidationData data
    ) {
        if (!(data.arg() instanceof List)) {
            return null;
        }
        if (data.containsPathToSchemas() == null) {
            return null;
        }
        if (data.containsPathToSchemas().size() < minContains) {
            throw new ValidationException(
                "Validation failed for minContains keyword in class="+data.schema().getClass()+
                " at pathToItem="+data.validationMetadata().pathToItem()+". Too few items validated to the contains schema."
            );
        }
        return null;
    }
}

File: samples/client/petstore/kotlin/src/main/kotlin/org/openapijsonschematools/client/schemas/validation/MinContainsValidator.kt
package org.openapijsonschematools.client.schemas.validation

import org.openapijsonschematools.client.exceptions.ValidationException

class MinContainsValidator : KeywordValidator {
    @Throws(ValidationException::class)
    override fun validate(
        data: ValidationData
    ): PathToSchemasMap? {
        val minContains: Int = data.schema.minContains ?: return null
        if (data.arg !is List<*>) {
            return null
        }
        if (data.containsPathToSchemas == null) {
            return null
        }
        if (data.containsPathToSchemas.size < minContains) {
            throw ValidationException("""
                |Validation failed for minContains keyword in class=${data.schema.javaClass} at
                | pathToItem=${data.validationMetadata.pathToItem}.
                | Too few items validated to the contains schema.
                """.trimMargin()
            )
        }
        return null
    }
}